esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -860,6 +860,29 @@ void EsdlServiceImpl::handleFinalRequest(IEspContext &context,
 
             soapmsg.append(req.str());
             soapmsg.append("</").append(tgtQueryName).append("></soap:Body>");
+
+            auto cfgGateways = tgtcfg->queryBranch("Gateways");
+            if (cfgGateways)
+            {
+                auto baseXpath = cfgGateways->queryProp("@transformTarget");
+                if (!isEmptyString(baseXpath))
+                {
+                    Owned<IPTree> gws = createPTree("gateways", 0);
+                    VStringBuffer tmpSoap("%s</soap:Envelope>", soapmsg.str());
+                    Owned<IPTree> soapTree = createPTreeFromXMLString(tmpSoap);
+                    StringBuffer xpath(baseXpath);
+
+                    EsdlBindingImpl::transformGatewaysConfig(tgtcfg, gws, "row");
+                    xpath.replaceString("{$query}", tgtQueryName);
+                    xpath.replaceString("{$method}", mthdef.queryMethodName());
+                    xpath.replaceString("{$service}", srvdef.queryName());
+                    xpath.replaceString("{$request}", mthdef.queryRequestType());
```

Comment:
FUTURE: Just a note that in the future we should create a common routine to do these replacements all in one call,  more efficiently.  For example pass it a IPropertyTree of the variables and replace them all in one pass.  We can then create the property object once for the entire transaction and use it all over the place.  The routine would be shared by this code and the ESDL transform code.  Not a big fan of tons of replaceString operations.

Reply:
Agree with the addition of a common routine. There is a version of StringBuffer with an efficient multiple string replacement function that could be a good addition. It works on a std::vector of pairs, where the first pair field is the text to be found and the second is the replacement text. In this context, we could populate the vector once and apply the replacements where needed.